scrapy/responsetypes.py

```diff
@@ -102,18 +123,61 @@ def from_body(self, body):
         else:
             return self.from_mimetype('text')
 
+    def _guess_content_type(self, headers=None, url=None, filename=None):
+        if headers and b'Content-Type' in headers:
+            return tuple(headers.getlist(b'Content-Type'))
+
+        if headers and b'Content-Disposition' in headers:
+            filename = headers.get(b'Content-Disposition').split(b';')[-1].split(b'=')[-1].strip(b'"\'').decode()
+        elif url:
+            filename = url
+
+        if filename:
+            mimetype, encoding = self.mimetypes.guess_type(filename)
+            if encoding:
+                return (f"application/{encoding}".encode(),)
+            else:
+                return (mimetype.encode(),)
```

Comment:
There's an [`AttributeError: 'NoneType' object has no attribute 'encode'`](https://github.com/scrapy/scrapy/pull/5204/checks?check_run_id=3279070482#step:5:2645) at this line, which is consistent with the following (value for `filename` taken from the test):
```python
In [1]: from mimetypes import MimeTypes

In [2]: mimetypes = MimeTypes()

In [3]: filename = '/tmp/pytest-of-runner/pytest-0/test_download0/tests.test_downloader_handlers/FileTestCase/test_download/18xzf8lw/temp^'

In [4]: mimetypes.guess_type(filename)
Out[4]: (None, None)
```